### ISSUE TITLE
WebAssembly Instrument and Tracing Technology

### ISSUE THREAD

**Issue by rrwinterton** [WebAssembly/design#1344]
**WebAssembly Instrument and Tracing Technology**
The WebAssembly Specification has the following prefixes defined in the Instruction Opcode section.  The following is a proposal to add an instruction and tracing technology prefix opcode to the list below:

**Opcode** | **Name** | **Description**
0xff | reserved | Reserved   for unknown future language evolution 🌌
0xfe | threads | Expected   to be used for atomics 🦄
0xfd | simd | Expected   to be used for SIMD 🦄
0xfc | misc | Miscellaneous   operations 🎳
**0xXX | ITT | Instrument   and Tracing Technology**

The intent for this opcode prefix byte is to identify the marking in the WebAssembly code to start tracing or instrumenting.  This prefix can be used to only with instructions specified to use this prefix (currently limited to only the NOP, similar to 0xfd is only allowed for SIMD).  
The prefix byte opcode can be added to the instruction up to 15 times prior to instruction it precedes.  The use of added the prefix multiple times it so uniquely identify an instance of the mark if there are multiple markers that are within the WebAssembly kernel to be traced.  For example assume the prefix byte opcode is 0xF9.  The marker could be added as 0xF901 byte code sequence which would be interpreted as a NOP instruction with a marking ID of 1.  A marker of 0xF9F901 byte would be a NOP instruction with a marker ID of 2.  

**Backend Implementation Specific Not part of the specification used as example of platform specific implementation**
This would allow the backend implementation of the machine targeted code to implement the instrument and tracing technology (ITT) targeted for the backend platform without altering the functionality of the code.  
For example Intel machine would translate the bytecode sequence of 0xF01 to a MARK_ID of 1 with the following instruction sequence.  A two byte prefix of 0x64, 0x67 followed by two NOP instructions and putting the MARK_ID 1 into the ebx register.

#define STATIC_TRACING_SSC_MARK( MARK_ID )     \
        __asm__ __volatile__ (          \
        "\n    movl $"#MARK_ID", %%ebx" \
        "\n    .byte 0x64, 0x67, 0x90, 0x90, 0x90"  \
        : : : "%ebx", "memory" ); 

This sequence is already used as an ITT event in Intel VTune, and Intel Software Developer Emulator as a way to start and stop tracing events.  Having the ability to mark different trace location allow developers to add multiple marks within the WebAssembly code without having to change the functionality of the code and still isolate a very specific location in the WebAssembly code for code isolation and optimization.

**Comment by tlively** [comment]
The mechanism of adding multiple prefixes to an opcode is inconsistent with how prefixes work in the rest of WebAssembly, because the (single) prefix is actually part of the specified opcode for an instruction. If you want to support up to 15 versions of the same instruction, it would be better for the instruction to take an immediate argument specifying which version it is or to have 15 different instructions.

At a higher level, can you explain how the semantics of these instructions would be specified? Would the spec list them as nops with non-normative notes on how engines could use them? Would there be any constraints or other standardization on what engines are allowed to do with them? Would they be useful at all on non-Intel platforms or on the Web?

**Comment by rrwinterton** [comment]
That is a good idea to have an immediate argument with a single ITT operation.  The immediate could be restricted to a single byte.  The spec would list them as a NOP as you stated.  An engine would be used to identify the marker for ITT tools to identify their own specific ITT instruction generated by the engine.  For non Intel platforms it can be used the exact same way as an ITT tool marker.   The need is not Intel specific.  In fact it could also be used for a marker in Chrome Dev Tools to start and stop traces of WASM in dev tools.

**Comment by backes** [comment]
This sounds like debugging information that would be ignored by most engines.

Would it work to add a custom section to the Wasm binary which basically encodes a list of `<offset, mark>` pairs?
I could imagine that this is a bit harder to maintain during linking and optimization passes like `wasm-opt`, but it would be way more flexible and would not have to be spec'ed. Which tool is supposed to generate the marks?

**Comment by rrwinterton** [comment]
Depends on what the engine does.  The final result that is needed for hardware and software instrumentation to work is the engine needs to generate the byte the byte sequence in platform specific code that is a two byte Intel hardware prefixed NOP with an identifier in the ebx register. 
 as shown here:  "0x64, 0x67, 0x90, 0x90, 0x90".  The method in how this is accomplished I am open for ideas.  The goal is to make it easy to place in sections of code that needs special attention for optimizations.

**Comment by jacob-abraham** [comment]
# Proposed new ITT opcode

This new opcode is proposed as a performance instrument and tracing technology instruction implemented in the style of other previous additions to WebAssembly like SIMD extensions. All opcodes in this subset fall under a prefix of 0xF9, and currently only has one opcode.  The ITT instructions should have no functional differences on the resultant code executed.

# Opcode

The hex opcode for this new instruction is the same as a regular NOP, 0x01. This is combined with the prefix to give an opcode of 0xF901. This instruction takes an immediate I32 value and uses it to set a mark id for execution tracing. The function type is [] -> [].

# Use

This new instruction would allow for fine tuned execution tracing. The first appearance of the instruction would signify the start of the execution trace and the second appearance of the instruction would signify the end of the trace. This is primarily for performance optimizations purposes.

# x86 Assembly example

An example of the platform targeted code for x86 assembly is shown below.
```asm
push rbx                  ; save the value of the rbx register
REX.W movq rbx,IMM        ; put the immediate value in the rbx register
0x64 0x67 0x90 0x90 0x90  ; execute a 5 byte sequence with 3 NOP's, specific to tracing technology such as Intel VTune
pop rbx                   ; restore the rbx register
```

# Other architectures

Other architectures, may implement as needed for the targeted specific ITT functionality.

**Comment by taralx** [comment]
Does the value have to be static, or could we take it from the stack (i.e. `[i32] -> []`)?

**Comment by kmiller68** [comment]
This feels like it would be better as an imported library function that might intrinsic to that sequence of Nops for X86 on implementations that support it. While we don't have any intrinsic functions currently, this instruction no defined operational semantics, so, IMO, there's no point bloating the instruction space for a new "Nop".

I could also imagine an implementation would emit something like WebKit's [SuperSampler](https://trac.webkit.org/browser/webkit/trunk/Source/JavaScriptCore/bytecode/SuperSampler.h), which just writes to some global boolean (or unsigned count) then has another thread poll if that global is set or not.

**Comment by binji** [comment]
> While we don't have any intrinsic functions currently,

Not only that, but we don't really have a good way to define intrinsic functions currently -- we didn't reserve any import names for internal use.

**Comment by kmiller68** [comment]
> > While we don't have any intrinsic functions currently,
> 
> Not only that, but we don't really have a good way to define intrinsic functions currently -- we didn't reserve any import names for internal use.

One option might be to just define the semantics and leave it to the host to select the name and how it gets imported. For example, the web we could standardize some name that other hosts are free to mirror. Alternatively, we could just add a WebAssembly.intrinsics object that has all the intrinsics hanging off it and users can import that. I don't know how other non-web hosts do their import resolutions so I don't know how practical this would be for them.

**Comment by binji** [comment]
@kmiller68 that could work. One issue (brought up in the await proposal) is that this model of intrinsics means you don't know whether it's an intrinsic until instantiation time. That's fine if you generate code at instantiation time, but not all VMs do that.

**Comment by kmiller68** [comment]
> @kmiller68 that could work. One issue (brought up in the await proposal) is that this model of intrinsics means you don't know whether it's an intrinsic until instantiation time. That's fine if you generate code at instantiation time, but not all VMs do that.

Fair point, I don't have a great answer for that. Perhaps that reason might justify some new way to specify a builtin-import, e.g. something other than a string.

Although, then you get into requiring conditional sections for checking the existence of the builtin... and fall back to the polyfill otherwise... TC-39 has taught me this is a very non-contentious topic... Fun times all around... 😬

**Comment by binji** [comment]
I added this to the June 23rd CG meeting agenda: https://github.com/WebAssembly/meetings/blob/master/2020/CG-06-23.md
